### PR TITLE
Watermark emission policy

### PR DESCRIPTION
**Pull request by mtopolnik**

### CODE REVIEW

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/WatermarkEmissionPolicy.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+package com.hazelcast.jet;
+
+import javax.annotation.Nonnull;
+import java.io.Serializable;
+
+/**
+ * A policy object that decides when when the watermark has advanced
+ * enough to emit a new watermark item.
+ */
+public interface WatermarkEmissionPolicy extends Serializable {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/WatermarkEmissionPolicy.java]: can be `@FunctionalInterface`

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/WatermarkEmissionPolicy.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+     * lastEmittedWm}.
+     */
+    boolean shouldEmit(long currentWm, long lastEmittedWm);
+
+    /**
+     * Returns a policy that ensures that each emitted watermark has a higher
+     * value than the last one.
+     */
+    @Nonnull
+    static WatermarkEmissionPolicy suppressDuplicates() {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/WatermarkEmissionPolicy.java]: is there a reason this is here, but not together with the rest of the policies? would it make sense to put all in a separate class?

Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/WatermarkEmissionPolicy.java]: This is supposed to serve as the "null" policy that just protects the basic invariant of watermark emission. But we could put all three in here.

Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/WatermarkEmissionPolicy.java]: I'd prefer them to all be in same place, preferably in this class ?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/InsertWatermarkP.java [hunk 3]
```diff
@@ -39,34 +39,39 @@
 public class InsertWatermarkP<T> extends AbstractProcessor {
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/processor/InsertWatermarkP.java]: This should also be renamed to `InsertWatermarksP` (plural)

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/WatermarkEmissionPolicy.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+     * value should be emitted, given the last emitted value {@code
+     * lastEmittedWm}.
+     */
+    boolean shouldEmit(long currentWm, long lastEmittedWm);
+
+    /**
+     * Returns a policy that ensures that each emitted watermark has a higher
+     * value than the last one.
+     */
+    @Nonnull
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/WatermarkEmissionPolicy.java]: We should say more clearly that this method is useful only in special cases: "Does minimum possible throttling causing lot of overhead from watermarks. Useful for edge cases and tests."

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/WatermarkEmissionPolicy.java (hunk 2 above)
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/WatermarkEmissionPolicy.java]: This is already done.

**Comment by cangencer** [comment]
verify

**Comment by mtopolnik** [comment]
verify
